Coveo TypeScript - Code standards (first draft) when developing for the Cloud Platform

MichelML opened the pull request:
Here it is, first draft of our written code standards.  

Suggested review method: 
Read the three standards (TypeScript, React/Redux, and CSS/Sass) directly from the [branch](https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide) first (not from the PR), take notes of what you would like to add/remove in each standards.   
  
Suggested things to focus on:  
- Rules to remove  
- Rules to add  
- Better code examples  
- Publishing strategy (should we keep this on GitHub or move it somewhere else?)

MichelML committed:
first removals after full read

MichelML committed:
add small change after second read

MichelML committed:
remove all eslint rules or jscs rules in styleguide

MichelML committed:
change style guide title

MichelML committed:
complete introduction

MichelML committed:
complete introduction

MichelML committed:
make sub header bigger

MichelML committed:
complete first draft of TypeScript styleguide
complete first draft of React/Redux styleguide
complete first draft of CSS/SASS styleguide

MichelML committed:
remove linters

MichelML committed:
fix broken links

MichelML committed:
fix broken links

MichelML committed:
make title the same in each standards readme

MichelML committed:
fix broken links in table of contents

MichelML committed:
remove obsolete folders

MichelML committed:
remove travis file

MichelML committed:
remove useless information in package.json

MichelML committed:
remove all scripts from original airbnb style guide

MichelML committed:
replace .sass for .scss

MichelML committed:
change style guide for standards term

stephr0y commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+
+* Use soft tabs (2 spaces) for indentation
+* Prefer dashes over camelCasing in class names.
+* Avoid using ID selectors as much as possible
```
I would also add, avoid using HTML tag selector as much as possible.

stephr0y commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+
+### Border
+
+Use `0` instead of `none` to specify that a style has no border.
```
Why?

stephr0y commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+* Use the `.scss` syntax, never the original `.sass` syntax
+* Order your regular CSS and `@include` declarations logically (see below)
+
+### Ordering of property declarations
```
Our sass is being formatted automatically (the order). I think that this section is useless and might be confusing.

stephr0y commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+
+Prefer dash-cased variable names (e.g. `$my-variable`) over camelCased or snake_cased variable names. 
+  
+For every properties for which you use pixels (`px`), you should use (or define it if it does not exist yet) a Sass variable. 
```
Color values as well.

stephr0y commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+
+Again: **never nest ID selectors!**
+
+If you must use an ID selector in the first place (and you should really try not to), they should never be nested. If you find yourself doing this, you need to revisit your markup, or figure out why such strong specificity is needed. If you are writing well formed HTML and CSS, you should **never** need to do this.
```
Same for `!important`

stephr0y commented on react_redux/README.md:
```diff
@@ -192,16 +196,16 @@
 ## Quotes
 
-  - Always use double quotes (`"`) for JSX attributes, but single quotes (`'`) for all other JS. eslint: [`jsx-quotes`](http://eslint.org/docs/rules/jsx-quotes)
+  - Always use double quotes (`'`) for JSX attributes, and all other JS. 
```
Double or single?

stephr0y replied:
I feel like it would not be a bad idea to change from single to double quote when in JSX

MichelML replied:
this is a typo, should be `Always use single quotes`

stephr0y commented on react_redux/README.md:
````diff
@@ -585,7 +563,7 @@
     export default Link;
     ```
 
````
We might want a working tsx example for the above code.

MichelML replied:
thanks @sroy3 for all your comments! will go through them soon enough

MichelML replied:
Would you leave these standards in a github repo @sroy3 ?

stephr0y replied:
The thing is in the above code, it says it is an example, but the Typescript way is very different than this. Also use Typescript in each of our project. It'd be weird to say things like "You forgot to type your variable here." when nothing is typed in the examples.

stephr0y reviewed (changes requested):
I would go over most code example and change them to typescript since this is what we use. Also some examples do not use some rules previously explained.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#es6-object-shorthand I believe we would actually write this:
```
const atom = {
  value: 1,
  addValue: (value) => atom.value + value,
};
```

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#arrows--one-arg-parens with Typescript, the parameter should be typed so parentheses have to be used.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#modules--no-wildcard We try to not have default export so this would not happen. Also you usually have to use the wildcard import with package for it to work.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#modules--prefer-default-export We do not want default exports.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#commas--dangling Currently we would have to fix our tslint for trailing commas to be accepted on function parameters.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#naming--filename-matches-export Again no default exports.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#test-spies should read "Name your spies with the name of the entity you want to spy on".

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#interface-no-i Pretty Typescript enforces prefixing interfaces with I. It is the case for react-vapor and log browser. 

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#variables--define-where-used I feel like it is easier to read and find const/let when they are at the begining. But because of the return, I would approve the way it is done in the example. 

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#comparison--shortcuts Why? It adds more code, and you sometimes have to add a check for `undefined`, `null` and/or `false` as well. The problem is more when checking for falsy values when a 0 or an empty string is valid.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#blocks--braces Depending on the context, the last bad example would be good.

dcarmichael1 commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+* Put a space before the opening brace `{` in rule declarations
+* In properties, put a space after, but not before, the `:` character.
+* Put closing braces `}` of rule declarations on a new line
+* Put blank lines between rule declarations
```
I would also add:
* No trailing whitespaces.
* Never use tabs.
* Never have the opening bracket `{` alone on its line.

dcarmichael1 commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+}
+#lol-no {
+  // ...
+}
```
What's wrong with the last one? (#lol-no)

GermainBergeron replied:
It uses an id

dcarmichael1 replied:
We never put any style on specific id?

dcarmichael1 commented on css_sass/README.md:
```diff
@@ -0,0 +1,311 @@
+If you must use an ID selector in the first place (and you should really try not to), they should never be nested. If you find yourself doing this, you need to revisit your markup, or figure out why such strong specificity is needed. If you are writing well formed HTML and CSS, you should **never** need to do this.
+
+**[⬆ back to top](#table-of-contents)**
+
```
Would be nice to talk about sass file architecture and how to bundle them up together in the global sass file.

dcarmichael1 commented on react_redux/README.md:
```diff
@@ -1,6 +1,9 @@
-# Airbnb React/JSX Style Guide
+# Coveo React/Redux - Cloud Platform Standards {
```
Why the `{` at the end?

rousselm4 replied:
@dcarmichael1 There is also a } at the end

GermainBergeron commented on README.md:
```diff
@@ -205,8 +184,8 @@ Other Style Guides
     };
```
Usually we use 
```
// good
const atom = {
  value: 1,
  addValue: (value: type) => atom.value + value,
  },
};
```

GermainBergeron commented on README.md:
````diff
@@ -381,46 +298,50 @@ Other Style Guides
+<a name="arrays--bracket-newline"></a><a name="4.3"></a>
+  - [4.3](#arrays--bracket-newline) Use line breaks after open and before close array brackets if an array has multiple lines
 
   ```javascript
````
When there's only so few values I prefer to inline them

```
const arr = [
  [0, 1], 
  [2, 3],
  [4, 5],
];

const objectInArray = [
  {id: 1},
  {id: 2},
];

const numberInArray = [1, 2];
```

GermainBergeron commented on README.md:
```diff
@@ -460,8 +381,9 @@ Other Style Guides
+  - [5.1](#destructuring--object) Use object destructuring when accessing and using multiple properties of an object. 
 
     > Why? Destructuring saves you from creating temporary references for those properties.
 
```
I'm not sure about the best here, it's not a style that we use often

GermainBergeron commented on README.md:
```diff
@@ -591,11 +510,8 @@ Other Style Guides
+  - [6.5](#strings--escaping) Do not unnecessarily escape characters in strings. 
 
     > Why? Backslashes harm readability, thus they should only be present when necessary.
 
```
`const foo = '\'this\' is "quoted"';`
In this particular case I would permit the use of string template to avoid escaping quotes

GermainBergeron commented on README.md:
````diff
@@ -964,7 +845,7 @@ Other Style Guides
     ```
````
I'd use braces + return instead of parenthesis here

```
// good
['get', 'post', 'put'].map(httpMethod => {
  return Object.prototype.hasOwnProperty.call(
    httpMagicObjectWithAVeryLongName, 
    httpMethod,
  )
});
```

MichelML replied:
it goes against the rule, i suggest we remove it or leave it as is...they propose to do the following if we use brackets here:  

``` 
// good
['get', 'post', 'put'].map(httpMethod => {
  return (
      Object.prototype.hasOwnProperty.call(
        httpMagicObjectWithAVeryLongName, 
        httpMethod,
      )
  );
});
```

GermainBergeron commented on README.md:
```diff
@@ -1155,35 +1036,12 @@ Other Style Guides
+  - [10.1](#modules--use-them) Always use modules (`import`/`export`) over a non-standard module system. You can always transpile to your preferred module system. **N.B. As of the last edit of our javascript styleguide, we are currently migrating towards `import`/`export`, but some of our projects use non-standard module systems for now.** 
 
     > Why? Modules are the future, let's start using the future now.
 
```
Defaults are bad, don't use them

GermainBergeron commented on README.md:
```diff
@@ -1280,7 +1138,7 @@ Other Style Guides
 
```
I prefer the bad to the good here. It's easier to import in other modules and you won't break them if you change the default export

GermainBergeron commented on README.md:
```diff
@@ -1579,41 +1376,8 @@ Other Style Guides
+  - [13.5](#variables--unary-increment-decrement) Avoid using unary increments and decrements (++, --). 
 
     > Why? Per the eslint documentation, unary increment and decrement statements are subject to automatic semicolon insertion and can cause silent errors with incrementing or decrementing values within an application. It is also more expressive to mutate your values with statements like `num += 1` instead of `num++` or `num ++`. Disallowing unary increment and decrement statements also prevents  ...
 
```
I'm not quite sure about this rule. I'm use count++ when I write my component tests to be able to increment a value and reorder my calls easily

```
let count = 0;
Utils.mockAjaxReceiveAt(200, someObject, count++);
Utils.mockAjaxReceiveAt(200, someOtherObject, count++);
Utils.mockAjaxReceiveAt(200, thirdObject, count++);
```

MichelML replied:
it's really nice for the case you present, but as a general prevention i like the reasoning of the doc that says: "Disallowing unary increment and decrement statements prevents you from pre-incrementing/pre-decrementing values unintentionally which can also cause unexpected behavior in your programs."

GermainBergeron commented on README.md:
````diff
@@ -1812,8 +1483,6 @@ Other Style Guides
-
     ```javascript
     // bad
     switch (foo) {
````
Can you remove the function declaration in case 3 and the class declaration in default? Both are bad examples IMO

GermainBergeron commented on README.md:
```diff
@@ -1962,46 +1621,17 @@ Other Style Guides
+        && doesItLookGoodWhenItBecomesThatLong()
+        && isThisReallyHappening()) {
+        thing1();
     }
```
I think we should choose between both styles (control statements at the start or at the end)

Personally I prefer at the start

GermainBergeron commented on README.md:
```diff
@@ -3080,314 +2577,129 @@ Other Style Guides
-included in all copies or substantial portions of the Software.
+  <a name="function-non-required-arguments"></a>
+  - [31.6](#function-non-required-arguments) Always put non required arguments last. 
+    > Why? If your function takes multiple arguments, you will only have to pass the required ones.
```
AFAIK this is enforced by Typescript

MichelML replied:
you're right removed

GermainBergeron commented on react_redux/README.md:
````diff
@@ -19,17 +22,18 @@
+  - If you have internal state and/or refs, prefer `class extends React.Component` over `React.createClass`. 
 
     ```jsx
     // bad
````
I think we should use `extends React.Component` over React.createClass and Function

MichelML replied:
this is what it says right?

GermainBergeron commented on react_redux/README.md:
````diff
@@ -78,9 +82,9 @@
+  - **Reference Naming**: Use PascalCase for React components and camelCase for their instances. 
 
     ```jsx
     // bad
````
```
// good
import {ReservationCard} from './ReservationCard';
```

GermainBergeron replied:
Same thing for other imports

GermainBergeron commented on react_redux/README.md:
````diff
@@ -464,7 +442,7 @@
     ```
 
-  - Bind event handlers for the render method in the constructor. eslint: [`react/jsx-no-bind`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md)
+  - Bind event handlers for the render method in the constructor. 
````
_I think_ we can use `const onClickDiv = () => {...}` and `onClick={this.onClickDiv}` without `bind`. Would need to test

GermainBergeron reviewed (commented):
I think we should review the javascript example to write them in typescript, with types.

In CSS/SASS it's quite different from what we use in the product right now and we should change the doc or the code accordingly

rousselm4 commented on react_redux/README.md:
```diff
@@ -19,17 +22,18 @@
-    - However, multiple [Stateless, or Pure, Components](https://facebook.github.io/react/docs/reusable-components.html#stateless-functions) are allowed per file. eslint: [`react/no-multi-comp`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-multi-comp.md#ignorestateless).
-  - Always use JSX syntax.
+    - However, multiple [Stateless, or Pure, Components](https://facebook.github.io/react/docs/reusable-components.html#stateless-functions) are allowed per file. 
+  - Always use JSX/TSX syntax.
```
I think we can remove JSX/

rousselm4 commented on react_redux/README.md:
```diff
@@ -611,25 +589,20 @@
 
 **[⬆ back to top](#table-of-contents)**
+
+# };
```
gg Très clean!

rousselm4 commented:
When we want to do something depending on the length of an array, we can proceed by many different ways. We should choose one and stick to this way. 

These are some way to do it:

1. `if (!!myArray.length) { ... }` & `if (!myArray.length) { ... }` with a boolean cast

2. `if (myArray.length > 0) { ... }` & `if (myArray.length < 1) { ... }` gt/lt operator

3. `if (myArray.length !== 0) { ... }` & `if (myArray.length === 0) { ... }` with equal/ not equal

4. `if (!_(myArray).isEmpty()) { ... }` & `if (_(myArray).isEmpty()) { ... }` with underscore (object way)

5. `if (!_.isEmpty(myArray)) { ... }` & `if (_.isEmpty(myArray)) { ... }` with underscore (functionnal way)

6. `if (myArray.length) { ... }` & `if (!myArray.length) { ... }` with implicit cast

Do you have any preferences?

stephr0y commented:
@rousselm4 In this context, definitely 6. It is simple and length here is predictable, also the only thing we are checking for is if there is something inside the array, not whether the length is a specific number. Even in spoken language, we would say "Does my array have a length", not "Is the length of my array something else than zero" or "Is the length of my array bigger than zero".

MichelML commented:
@rousselm4 along the same lines as @sroy3 but would go for `1` since it is more explicit, I think the Airbnb style guide recommends this (casting other types into boolean for if/else or conditional statements)

MichelML committed:
refactor all examples in typescript

MichelML committed:
refactor function for arrow functions or class methods

MichelML commented:
@GermainBergeron can you provide some more information on the way you would like the SASS/CSS to be presented?

MichelML committed:
apply reviews of everyone

MichelML commented:
ok so i think i applied most of your comments guys @rousselm4 @sroy3 @GermainBergeron , unless I specified otherwise.

Here are additional things I did:  

- removed computed property rule for object
- removed object destructuring of a function argument function({arg1, arg2}) {} 
- add interfaces everywhere  
- removed functions—in-blocks rule (not essential)
- removed rule ‘functions--default-side-effects’ (not essential, repetitive)
- removed rule functions--defaults-last (enforced by TS)
- removed #functions--signature-spacing (mostly not related to TS and the way we code)
- arrows--one-arg-parens —> changed for always use parentheses
- remove rules related to `import *.scss ` 
- logical operator should be placed at beginning of line
- naming--filename-matches-export — removed
- removed naming—self-this 
- removed - **Higher-order Component Naming**: 
- refactored functions in the format const myFunc = () => {} instead of function() {}

feel free to provide additional comments, after that i'll focus on automating the standards with a proper tslint and i'll look for additional tools.  
  
Some things you should have a look at: 1) do the examples in TypeScript are good for you? You can check for typing style mostly and the way i refactored functions.

MichelML committed:
adjust style in examples

MichelML committed:
refactor introduction

MichelML committed:
refactor introduction

stephr0y commented:
For https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#arrays--bracket-newline have we talked about it before? I feel like it is the first time I see it. I feel that
```
const objectInArray: Interface[] = [{
  id: 1,
}, {
  id: 2,
}];
```
is much cleaner than
```
const objectInArray: Interface[] = [
  {
    id: 1,
  },
  {
    id: 2,
  },
];
```
All other examples are fine though.

I do not know how I feel about https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#destructuring--object I like the concept, but I do not think I have seen an example where it is worth while. For your example, 
```
const getFullName = (user: User): string => {
  return `${user.firstName} ${user.lastName}`;
}
```
is better because it does not create unnecessary variables and shorten the code. I could see this work in some case like this:
```
const getFullName = (user: User): string => {
  let { firstName, lastName } = user;
  firtName += ' suffix';
  lastName = lastName.length > 5 ? lastName : `${lastName} ${lastName}`;
  return `${firstName} ${lastName}`;
}
```

For https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#functions--mutate-params, we would probably just do `const key: number =  obj.key || 1;` since `obj` is mandatory. Or `const key: number = obj && obj.key || 1;` if it isn't. Also it skips from 7.5 to 7.11 for this section.

Isn't https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#functions--reassign-params the same as https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#es6-default-parameters ?

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#functions--signature-invocation-indentation trailing comma isn't working yet with our settings. It would be nice to investigate so that it works though.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#modules--no-duplicate-imports We usually have imports on one line if line is too long. The "Optimize imports" in WebStorm does this for us. Might need to specify here https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#modules--multiline-imports-over-newlines too.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#comments--actionitems and https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#comments--todo are pretty much the same thing (also where is 18.5?) they could be merge together.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#whitespace--after-blocks shouldn't brackets of empty blocks be on the same line?

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#whitespace--max-len after refactoring V2, I've found many readability issues with line breaks. I would add that the line should be broken at a meaningful position and that if it barely breaks the 140 characters to just leave it as is instead of breaking the line and making it harder to read.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#naming--PascalCase I feel like the bad and the good examples aren't really showing the same thing. The bad one should also have a class.

https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#review-daily Typo here "loosing" should be "losing".

rousselm4 commented on css_sass/README.md:
```diff
@@ -0,0 +1,270 @@
+
+1. [General rules](#general-rules)
+1. [Terminology](#terminology)
+    - [Rule Declaration](#rule-declaration)
```
C'est un détail, mais parfois c'est en Title Case et d'autre fois non. Ça serait bien de juste tout uniformiser les titres.

GermainBergeron commented on README.md:
````diff
@@ -254,194 +202,126 @@ Other Style Guides
-  - [4.5](#arrays--callback-return) Use return statements in array method callbacks. It's ok to omit the return if the function body consists of a single statement following [8.2](#arrows--implicit-return). eslint: [`array-callback-return`](http://eslint.org/docs/rules/array-callback-return)
-
-    ```javascript
+    ```typescript
````
I think we should always put the "bad" section first, or at least always have the same order

MichelML replied:
it is actually the case, except for this specific case i think, i added an example

GermainBergeron commented on README.md:
```diff
@@ -254,194 +202,126 @@ Other Style Guides
+    inbox.filter((msg: string): boolean => msg.subject === 'Mockingbird'
+        ? msg.author === 'Harper Lee'
+        : false
+    );
```
inbox.filter((msg: string): boolean => msg.subject === 'Mockingbird' && msg.author === 'Harper Lee');

MichelML replied:
it's for the example (if else), left as is

GermainBergeron commented on README.md:
```diff
@@ -516,337 +390,221 @@ Other Style Guides
-      const key = Object.prototype.hasOwnProperty.call(obj, 'key') ? obj.key : 1;
-    }
+    const f2 = (obj: Interface) => {
+      const key: number = Object.prototype.hasOwnProperty.call(obj, 'key') ? obj.key : 1;
```
I'm not sure we use `hasOwnProperty` in our code base. Maybe something like 

`const key = obj.key || 1;` 
or 
`f2 = (obj: Interface = {key: 1}) => {...`

MichelML replied:
the shortest would be to use `_.has`, will implement (and less error prone imho)

GermainBergeron commented on README.md:
```diff
@@ -1073,178 +811,166 @@ Other Style Guides
   <a name="modules--no-export-from-import"></a><a name="10.3"></a>
   - [10.3](#modules--no-export-from-import) And do not export directly from an import.
 
     > Why? Although the one-liner is concise, having one clear way to import and one clear way to export makes things consistent.
```
I actually prefer the `export { ... } from 'x';` syntax in the case of the index.ts files (like in react-vapor). What do you think? @sroy3 @MichelML @pochretien @rousselm4

GermainBergeron commented on README.md:
```diff
@@ -1073,178 +811,166 @@ Other Style Guides
-    // good
-    import foo, { named1, named2 } from 'foo';
+    // bad
+    import { named1, named2 } from 'foo';
```
We do this nearly everywhere and only wraps when there's a lot of import

MichelML replied:
thinking in an automation mindset, i think we should try to simplify things as much, adding some rule like 'if more than two imports' would complexify stuff...i would put it at a strict do or don't...unless our automation tools allow us otherwise...

GermainBergeron commented on README.md:
```diff
@@ -2337,34 +1711,14 @@ Other Style Guides
     };
```
```
const obj: Interface = {
  foo: () => {},
  bar: () => {},
}
```

MichelML replied:
changed

GermainBergeron commented on README.md:
```diff
@@ -2418,43 +1772,43 @@ Other Style Guides
 
   <a name="whitespace--in-braces"></a><a name="18.11"></a>
-  - [19.11](#whitespace--in-braces) Add spaces inside curly braces. eslint: [`object-curly-spacing`](http://eslint.org/docs/rules/object-curly-spacing.html) jscs: [`requireSpacesInsideObjectBrackets`](http://jscs.info/rule/requireSpacesInsideObjectBrackets)
+  - [19.11](#whitespace--in-braces) Add spaces inside curly braces. 
```
Any reason for this? It seems weird to do it for curly braces but not brackets and parentheses

MichelML replied:
agree! changed

GermainBergeron commented on README.md:
```diff
@@ -3080,314 +2237,114 @@ Other Style Guides
-  - [View Contributors](https://github.com/airbnb/javascript/graphs/contributors)
+  <a name="interface-no-i"></a>
+  - [31.2](#interface-no-i) Do not prefix your interfaces with `I`.  
+    > Why? Unconventional. Important interfaces (like `Window`, `Document`) are usually not defined with the `I` prefix.
```
We do otherwise in react-vapor

MichelML replied:
we should change it in react-vapor...this is the shortest path to apply our standards everywhere

GermainBergeron commented on react_redux/README.md:
````diff
@@ -0,0 +1,540 @@
+
+    ```typescript
+    // bad
+    import { Footer } from './Footer/Footer';
````
Currently we do this in react-vapor and I prefer it to the "good" in this section. It's harder to find a file if you have 200 index.tsx in your project

MichelML replied:
changed

GermainBergeron commented on react_redux/README.md:
````diff
@@ -0,0 +1,540 @@
+    }
+    ```
+
+  - Bind event handlers for the render method in the constructor. 
````
Use arrow function instead of bind

MichelML replied:
changed

GermainBergeron commented on react_redux/README.md:
```diff
@@ -0,0 +1,540 @@
+      }
+
+      render(): JSX.Element {
+        return <div onClick={this.onClickDiv} />;
```
onClick={() => this.onClickDiv()}

GermainBergeron commented on react_redux/README.md:
````diff
@@ -0,0 +1,540 @@
+    ```
+
+  - Do not use underscore prefix for internal methods of a React component.
+    > Why? Underscore prefixes are sometimes used as a convention in other languages to denote privacy. But, unlike those languages, there is no native support for privacy in JavaScript, everything is public. Regardless of your intentions, adding underscore prefixes to your properties does not actually make them private, and any property (underscore-prefixed or not) should be treated as being pub ...
````
Here you can add a word about `private` method in typescript?

MichelML replied:
done

GermainBergeron commented on react_redux/README.md:
```diff
@@ -0,0 +1,540 @@
+
+    // good
+    class extends React.Component<Props, void> {
+      onClickSubmit() {
```
`private onClickSubmit() {`

MichelML replied:
done

MichelML committed:
apply germain last reviews

MichelML commented:
@sroy3 for https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#arrays--bracket-newline , i felt that way before too, but i've come to find the one being in the standards now: 

```
const objectInArray: Interface[] = [
  {
    id: 1,
  },
  {
    id: 2,
  },
];
```
mainly because everything is symmetric, and thus easier to copy, cut, and paste (duplicate, reorder, refactor, change places)  things when we write it this way.

MichelML commented:
@sroy3 for https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#destructuring--object , i agree with you that it's definitely overkill for some cases, but where it makes sens it can definitly makes things less verbose. Plus, (not 100% sure), but suppose you have this in TS:
```
    const getFullName = (user: User): string => {
      const { firstName, lastName } = user;
      return `${firstName} ${lastName}`;
    }
```
i think it is actually compiled to something like

```
    var getFullName = (use) => {
      return `${user.firstName} ${user.lastName}`;
    }
```
in js (they remain accessors even if they are destructured)

MichelML commented:
for https://github.com/coveo/typescript-cloudplatform/tree/coveo-javascript-styleguide#functions--signature-invocation-indentation unless i'm wrong i think it's working now, we'll make it work anyway :D

MichelML committed:
apply steph reviews
reorder sections
add being reviewed section
add rule in reviewing section

MichelML commented:
@sroy3 i think i've applied most of your last comments either through the reviews of Germain or through yours...the next step will be to automate this standards which will probably shake things a little again

MichelML commented:
@rousselm4 @sroy3 @GermainBergeron @pochretien @vseguin I added a section "Being reviewed", feel free to have a look if you want! it's short

MichelML commented:
also, all sections have been properly numbered and ordered

stephr0y commented:
@MichelML for destructuring objects, I can keep it, but I would add that it should be used to help with short hand properties (by renaming) or by accessing deeply nested variables (or simply making it less long). In other words, its use should result in less code, not more. Maybe change the example for something like: 
```
const getFullName = (user: User): string => {
    const { firstName, lastName } = user.personnalInformation;
    return `${firstName} ${lastName}`;
}
```
This way it shows the power of destructuring a little better.

MichelML commented:
i'll apply your review and merge @sroy3 , thanks!

MichelML committed:
apply steph last review

MichelML commented:
last request applied
